### PR TITLE
Staff Payment and Office Address Bugs

### PR DESCRIPTION
**Pull request by cameron-freshworks**
*Issue #:* /bcgov/entity#7416

*Description of changes:*

* Include Staff payment in filing and draft parsing.

* Added Staff payment to alteration filing build.

* Created method to parse staff payment into store for both alterations and corrections.

* Fixed Office addresses bug where it wouldn't initialize the data after changing views, would only update the component when the Store values changed, which only happened once in Alterations.

* verified resolution dates are working as expected.

* Minor unit test updates to include the additional event.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "business-edit-ui",
-  "version": "0.6.6",
+  "version": "0.6.7",
```
Comment (severinbeauvais) [inline package.json:3]: This will be 0.6.8 now that Andre's PR is merged.

File: src/components/YourCompany/OfficeAddresses.vue [hunk 2]
```diff
@@ -624,7 +629,7 @@ export default class OfficeAddresses extends Mixins(CommonMixin) {
    * sets local properties and emits state events.
    */
   @Watch('getOfficeAddresses', { deep: true })
```
Comment (severinbeauvais) [inline src/components/YourCompany/OfficeAddresses.vue:631]: Why not use `immediate: true`  to fire this method up front?

File: src/mixins/filing-template-mixin.ts [hunk 3]
```diff
@@ -283,6 +283,29 @@ export default class FilingTemplateMixin extends Mixins(DateMixin) {
     // Include name request info when applicable
     if (this.hasNewNr || this.hasBusinessNameChanged) filing.alteration.nameRequest = { ...this.getNameRequest }
 
+    // Populate Staff Payment according to payment option
```
Comment (severinbeauvais) [inline src/mixins/filing-template-mixin.ts:286]: This looks identical to the code for corrections... could this use a new helper method?

File: src/mixins/filing-template-mixin.ts [hunk 4]
```diff
@@ -611,4 +601,48 @@ export default class FilingTemplateMixin extends Mixins(DateMixin) {
     }
     this.setBusinessContact(contactPoint)
   }
+
+  /** Parse Staff Payment data into store.
+   * @param filing The alteration or correction filing to parse.
+   */
+  storeStaffPayment (filing: AlterationFilingIF | CorrectionFilingIF): void {
```
Comment (severinbeauvais) [inline src/mixins/filing-template-mixin.ts:608]: Private.

File: src/components/YourCompany/OfficeAddresses.vue (hunk 2 above)
Comment (cameron-freshworks) [reply src/components/YourCompany/OfficeAddresses.vue:631]: It's fired up front already. The issue is during the re-mount ie changing to review and back to company changes. 
Would immediate fire it on a re-mount?

File: package.json (hunk 1 above)
Comment (cameron-freshworks) [reply package.json:3]: oh nice, thanks for the heads up. Will update!

File: tests/unit/OfficeAddresses.spec.ts [hunk 5]
```diff
@@ -957,7 +957,7 @@ describe('actions and events', () => {
 
     // verify initial events
     const haveChanges = wrapper.emitted('haveChanges')
-    expect(haveChanges.length).toBe(1)
+    expect(haveChanges.length).toBe(2)
```
Comment (severinbeauvais) [inline tests/unit/OfficeAddresses.spec.ts:960]: Why are there 2 initial events -- what are they?

File: src/mixins/filing-template-mixin.ts (hunk 3 above)
Comment (cameron-freshworks) [reply src/mixins/filing-template-mixin.ts:286]: It is! Yes you bet, i did the same for the parse, should have done this for the build 😁

File: src/mixins/filing-template-mixin.ts (hunk 4 above)
Comment (cameron-freshworks) [reply src/mixins/filing-template-mixin.ts:608]: indeed, will do!

File: tests/unit/OfficeAddresses.spec.ts (hunk 5 above)
Comment (cameron-freshworks) [reply tests/unit/OfficeAddresses.spec.ts:960]: Mounted lifecycle and the Watcher

File: src/components/YourCompany/OfficeAddresses.vue (hunk 2 above)
Comment (severinbeauvais) [reply src/components/YourCompany/OfficeAddresses.vue:631]: There is no such thing as a remount. The component is actually being re-loaded/re-rendered.

(This is part of why NR was nasty -- the components are always (mostly) rendered, and conditionally displayed. In that case you have to catch the display event and init there.)

Comment (severinbeauvais) [reply src/components/YourCompany/OfficeAddresses.vue:631]: Short answer: yes, when the component is reloaded, all immediate watchers should fire.

What is firing this up front already? Why doesn't that do what it's supposed to with the data?

File: tests/unit/OfficeAddresses.spec.ts (hunk 5 above)
Comment (severinbeauvais) [reply tests/unit/OfficeAddresses.spec.ts:960]: If the watcher fires when `getOfficeAddresses` changes, why do you need to call it from mounted() ?

File: src/components/YourCompany/OfficeAddresses.vue (hunk 2 above)
Comment (cameron-freshworks) [reply src/components/YourCompany/OfficeAddresses.vue:631]: > There is no such thing as a remount. The component is actually being re-loaded/re-rendered.
> 
> (This is part of why NR was nasty -- the components are always (mostly) rendered, and conditionally displayed. In that case you have to catch the display event and init there.)

I'm not sure what you want to call it, re-mounting seemed most accuratre. The component itself is mounted every time you change to Review and back to the components.

It fires up front because the store values change with regards to the office addresses, every time a filing is loaded. Why it works initially but doesn't when it 'mounts again'. Because there is no changes to the store value at that point.

File: tests/unit/OfficeAddresses.spec.ts (hunk 5 above)
Comment (severinbeauvais) [reply tests/unit/OfficeAddresses.spec.ts:960]: (Not that I've never done it before, but it's a code smell when a component emits the same event multiple times on init / when nothing has happened yet.)

File: src/components/YourCompany/OfficeAddresses.vue (hunk 2 above)
Comment (severinbeauvais) [reply src/components/YourCompany/OfficeAddresses.vue:631]: > I'm not sure what you want to call it, re-mounting seemed most accuratre. The component itself is mounted every time you change to Review and back to the components.

Haha, I agree with you there. Let's call it "remounted" then.

My point is that this component is destroyed between mounted calls. It's not just dormant and re-initialized.

Comment (severinbeauvais) [reply src/components/YourCompany/OfficeAddresses.vue:631]: What if this component wasn't rendered (mounted!) until the data is loaded...?

Comment (cameron-freshworks) [reply src/components/YourCompany/OfficeAddresses.vue:631]: Totally, the issue was purely that the initialization was based on a watcher (office address store data) which only changes once in an Alteration. Not an issue in corrections because the single view, the component is never destroyed.

I think why this isn't working like our other components, is the value fed into the Base Address Components isn't exactly reactive, it is computed values which the component is internally doing, and in this case the trigger for that is a change.

I'll try `immediate` though. Would much prefer that as well!

File: tests/unit/OfficeAddresses.spec.ts (hunk 5 above)
Comment (cameron-freshworks) [reply tests/unit/OfficeAddresses.spec.ts:960]: I agree, see the comment above 😁 I can elaborate more though, if need be,

File: src/components/YourCompany/OfficeAddresses.vue (hunk 2 above)
Comment (cameron-freshworks) [reply src/components/YourCompany/OfficeAddresses.vue:631]: immediate works great, thanks! 

I didn't know this option did this, assumed it tied into `created` hook, this is really good to know 👍

File: package.json (hunk 1 above)
Comment (cameron-freshworks) [reply package.json]: Updated

File: tests/unit/OfficeAddresses.spec.ts (hunk 5 above)
Comment (andrepestana-aot) [reply tests/unit/OfficeAddresses.spec.ts:960]: Yeah, just to give my 2 cents: maybe checking in the watch if the value has really changed may resolve it if it's because of the mount that triggers it. I'm not sure but there may be any option to discard it.

File: src/mixins/filing-template-mixin.ts (hunk 3 above)
Comment (cameron-freshworks) [reply src/mixins/filing-template-mixin.ts]: Updated

File: src/mixins/filing-template-mixin.ts (hunk 4 above)
Comment (cameron-freshworks) [reply src/mixins/filing-template-mixin.ts]: updated

File: tests/unit/OfficeAddresses.spec.ts (hunk 5 above)
Comment (cameron-freshworks) [reply tests/unit/OfficeAddresses.spec.ts:960]: With the inclusion of `immediate` the changes still stand as it will call the method an additional time.
When the component is mounted and when the watcher is triggered, ie any changes to the office addresses.

Comment (severinbeauvais) [reply tests/unit/OfficeAddresses.spec.ts:960]: A small comment where the component emits the event, to describe the init sequence (and thus why there are multiple events), would be great! ❤️

Comment (cameron-freshworks) [reply tests/unit/OfficeAddresses.spec.ts:960]: Will do 👍

File: src/mixins/filing-template-mixin.ts [hunk 6]
```diff
@@ -611,4 +561,78 @@ export default class FilingTemplateMixin extends Mixins(DateMixin) {
...
+      case StaffPaymentOptions.NO_FEE:
+        filing.header.waiveFees = true
+        filing.header.priority = false
+        break
+
+      case StaffPaymentOptions.NONE: // should never happen
+        break
+    }
+
+    return filing
```
Comment (severinbeauvais) [inline src/mixins/filing-template-mixin.ts:592]: Theoretically this isn't needed, because the code above changes the original object. (If you go this route, you'd have to update lines 170 and 267.)

Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:592]: The object is passed by reference I think (like a pointer to an object in C).

Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:592]: Or maybe not.
https://stackoverflow.com/questions/13104494/does-javascript-pass-by-reference

Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:592]: https://www.geeksforgeeks.org/pass-by-value-and-pass-by-reference-in-javascript/

I'm going to assume the above are correct and resolve this conversation with no change required ;)

File: src/components/YourCompany/OfficeAddresses.vue [hunk 7]
```diff
@@ -620,15 +620,16 @@ export default class OfficeAddresses extends Mixins(CommonMixin) {
   }
 
   /**
-   * When stored office addresses change (ie, when data is loaded/updated/reset),
+   * When the component is mounted or stored office addresses change (ie, when data is loaded/updated/reset),
    * sets local properties and emits state events.
```
Comment (severinbeauvais) [inline src/components/YourCompany/OfficeAddresses.vue:624]: Can you also explain why this component emits this event twice on init?

File: src/mixins/filing-template-mixin.ts (hunk 6 above)
Comment (cameron-freshworks) [reply src/mixins/filing-template-mixin.ts:592]: Hmm.. i thought it would be. 
If it isn't it would be cleaner and preferable. I will try!

**Review by severinbeauvais — APPROVED** [review]
Ready to merge when you are.

File: src/components/YourCompany/OfficeAddresses.vue (hunk 7 above)
Comment (cameron-freshworks) [reply src/components/YourCompany/OfficeAddresses.vue:624]: Doesn't this already explain that? 
How would you word it?

**Review by andrepestana-aot — APPROVED** [review]
👍 Looks good to me!

File: src/components/YourCompany/OfficeAddresses.vue (hunk 7 above)
Comment (cameron-freshworks) [reply src/components/YourCompany/OfficeAddresses.vue:624]: It only emits once on init (mounted due to immediate) and it emits again once the data changes (in our case, when the filing is loaded, and any subsequent changes). 

hehe i can write that.. though it seems long winded. I need your 'comment' expertise here!

File: src/mixins/filing-template-mixin.ts (hunk 6 above)
Comment (cameron-freshworks) [reply src/mixins/filing-template-mixin.ts:592]: Yea.. works good! I was worried the filing would return without the changes.
Good clean up, thanks!

File: src/components/YourCompany/OfficeAddresses.vue (hunk 7 above)
Comment (severinbeauvais) [reply src/components/YourCompany/OfficeAddresses.vue:624]: I'm not looking for "what" it's doing (ie, immediate + watching) but "why".

> It fires up front because the store values change with regards to the office addresses, every time a filing is loaded. Why it works initially but doesn't when it 'mounts again'. Because there is no changes to the store value at that point.

This sort of explained it for me but I will soon forget it, so hoping to get this into a code comment.

I guess the gist is that this component loads/inits/mounts before the data is loaded, and also after. Or something about the reactivity not set up to work as intended. I dunno.

Comment (cameron-freshworks) [reply src/components/YourCompany/OfficeAddresses.vue:624]: I'll try my best to do so without getting too wordy :)

Comment (severinbeauvais) [reply src/components/YourCompany/OfficeAddresses.vue:624]: > It only emits once on init (mounted due to immediate) and it emits again once the data changes (in our case, when the filing is loaded, and any subsequent changes).

Yeah, this.

But now I'm wondering why this component is displayed before it had data 😛  (IMHO it shouldn't.) But that can be reworked another day.

File: src/mixins/filing-template-mixin.ts (hunk 6 above)
Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts]: OK so now I'm confused why it works 🤷

File: src/components/YourCompany/OfficeAddresses.vue (hunk 7 above)
Comment (cameron-freshworks) [reply src/components/YourCompany/OfficeAddresses.vue:624]: It's an internal issue in the component i think.
The component computes the values which it then feeds into the Base Address Sub component. This was only being called when the Store values changed. 
I think if the Base Addres sub components were utilizing the store getter directly it wouldn't have been an issue, but the Parent component is doing a local computation of sorts.

File: src/mixins/filing-template-mixin.ts (hunk 6 above)
Comment (cameron-freshworks) [reply src/mixins/filing-template-mixin.ts]: Like you said, the javascript reference :) 

Tbh i dunno either, it does mutate the original values though.

Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts]: Maybe this?
https://dev.to/bytebodger/a-gotcha-of-javascript-s-pass-by-reference-1afm

Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts]: TLDR
